src/components/Elements.tsx

```diff
@@ -101,74 +87,49 @@ interface PrivateElementsProps {
  */
 export const Elements: FunctionComponent<ElementsProps> = ({
   stripe: rawStripeProp,
-  options,
+  options: optionsProp,
   children,
 }: PrivateElementsProps) => {
-  const final = React.useRef(false);
-  const isMounted = React.useRef(true);
-  const parsed = React.useMemo(() => parseStripeProp(rawStripeProp), [
-    rawStripeProp,
-  ]);
-  const [ctx, setContext] = React.useState<ElementsContextValue>(() => ({
-    stripe: null,
-    elements: null,
-  }));
-
-  const prevStripe = usePrevious(rawStripeProp);
-  const prevOptions = usePrevious(options);
-  if (prevStripe !== null) {
-    if (prevStripe !== rawStripeProp) {
+  const [inputs, setInputs] = React.useState({ rawStripe: rawStripeProp, options: optionsProp })
+  const { rawStripe, options } = inputs
+  React.useEffect(() => {
+    const hasRawStripeChanged = rawStripe !== rawStripeProp
+    const hasOptionsChanged = !isEqual(options, optionsProp)
+    const canUpdate = rawStripe === null
+
+    if (hasRawStripeChanged && !canUpdate) {
       console.warn(
         'Unsupported prop change on Elements: You cannot change the `stripe` prop after setting it.'
       );
     }
-    if (!isEqual(options, prevOptions)) {
+
+    if (hasOptionsChanged && !canUpdate) {
       console.warn(
         'Unsupported prop change on Elements: You cannot change the `options` prop after setting the `stripe` prop.'
       );
     }
-  }
 
-  if (!final.current) {
-    if (parsed.tag === 'sync') {
-      final.current = true;
-      setContext({
-        stripe: parsed.stripe,
-        elements: parsed.stripe.elements(options),
-      });
-    }
+    if (hasRawStripeChanged && canUpdate) setInputs({ rawStripe: rawStripeProp, options: optionsProp })
+  }, [rawStripe, options, rawStripeProp, optionsProp])
 
-    if (parsed.tag === 'async') {
-      final.current = true;
-      parsed.stripePromise.then((stripe) => {
-        if (stripe && isMounted.current) {
-          // Only update Elements context if the component is still mounted
-          // and stripe is not null. We allow stripe to be null to make
-          // handling SSR easier.
-          setContext({
-            stripe,
-            elements: stripe.elements(options),
-          });
-        }
-      });
-    }
-  }
+  const maybeStripe = usePromiseResolver(rawStripe)
+  const stripe = validateStripe(maybeStripe)
+  const [ctx, setContext] = React.useState<ElementsContextValue>(() => createElementsContext(null));
```

Comment:
blocking: I think this will cause a subtle change in behavior that may break existing integrations. Consider someone using React Stripe.js like this:

```jsx
// Synchronously create `Stripe` instance at module level
const stripe = Stripe('pk_123')

const CheckoutPage = () => {
  // Pass already instantiated `Stripe` instance to `Elements` provider
  return (
    <Elements stripe={stripe}>
      <CheckoutForm />
    </Elements>
  )
}

const CheckoutForm = () => {
  // Integration assumes `stripe` is not `null` on first render
  const stripe = useStripe();
  const [paymentRequest] = useState(() => stripe.paymentRequest(options))

  // ...etc.
}
```
This integration expects that if a valid `Stripe` instance is passed to an `Elements` provider, that `Stripe` instance is available via `useStripe` in a child component when the child component first renders. With the current behavior in this PR, the child component will be rendered twice and `useStripe` will return `null` during the first render.

Here's a failing test for this:

```jsx
it('exposes a passed Stripe instance immediately to children calling useStripe', () => {
  const TestChild = () => {
    const stripe = useStripe();

    if (stripe === null) {
      throw new Error('TestChild rendered with null stripe instance');
    }

    return null;
  };

  expect(() => {
    mount(
      <Elements stripe={stripe}>
        <TestChild />
      </Elements>
    );
  }).not.toThrow();
});

```

Reply:
I also thought about this. But declined by 2 reasons:

- The interface defines nullable values.
- The existing implementation also doesn't guarantee stripe instance on mount. As no one guarantees that the state will be updated before the first render of children ;-)

Anyway it's easy to add.

Reply:
Confirming - the existing code will provide empty values on the first render as well.